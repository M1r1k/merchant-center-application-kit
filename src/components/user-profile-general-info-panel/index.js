export { default } from './user-profile-general-info-panel';
