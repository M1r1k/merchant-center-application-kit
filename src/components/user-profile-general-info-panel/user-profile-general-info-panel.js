import React from 'react';
import PropTypes from 'prop-types';
import { compose, withProps } from 'recompose';
import { injectIntl, intlShape } from 'react-intl';
import { Field } from 'redux-form';
import ThrottledField from '@commercetools-local/core/components/fields/throttled-field';
import wrapInputForReduxForm from '@commercetools-local/core/components/input-wrapper';
import CollapsiblePanel from '@commercetools-local/ui-kit/panels/collapsible-panel';
import FormBox from '@commercetools-local/core/components/form-box';
import LabelField from '@commercetools-local/core/components/fields/label-field';
import messages from './messages';

const Input = compose(
  wrapInputForReduxForm,
  withProps(ownerProps => ({
    // the ThrottledField component only passes the DOM event as param when
    // calling the onChange handler but redux-form expects the changed value
    onChange: ({ target }) => ownerProps.onChange(target.value),
    // we need to pass onEnter here so that ThrottledField aborts pending
    // updates and calls onChange with the current value when hitting "Enter"
    // We don't add some custom implementation here because redux-form already
    // listens to the onKeyDown events on the field and submits the form when
    // "Enter" is pressed
    onEnter: () => {},
  }))
)(ThrottledField);

const shouldShowError = meta =>
  (meta.hasSubmitFailed || meta.touched) && meta.invalid;

export const UserProfileGeneralInfoPanel = props => (
  <CollapsiblePanel label={props.intl.formatMessage(messages.title)}>
    <FormBox>
      <LabelField
        title={props.intl.formatMessage(messages.firstName)}
        isRequired={true}
      />
      <Field
        name="firstName"
        component={Input}
        shouldDisplayTooltip={shouldShowError}
        submitFailed={props.hasSubmitFailed}
      />
    </FormBox>
    <FormBox>
      <LabelField
        title={props.intl.formatMessage(messages.lastName)}
        isRequired={true}
      />
      <Field
        name="lastName"
        component={Input}
        shouldDisplayTooltip={shouldShowError}
        submitFailed={props.hasSubmitFailed}
      />
    </FormBox>
    <FormBox>
      <LabelField title={props.intl.formatMessage(messages.email)} />
      <Field name="email" component={Input} disabled={true} />
    </FormBox>
  </CollapsiblePanel>
);
UserProfileGeneralInfoPanel.displayName = 'UserProfileGeneralInfoPanel';
UserProfileGeneralInfoPanel.propTypes = {
  hasSubmitFailed: PropTypes.bool,
  intl: intlShape.isRequired,
};

export default injectIntl(UserProfileGeneralInfoPanel);
