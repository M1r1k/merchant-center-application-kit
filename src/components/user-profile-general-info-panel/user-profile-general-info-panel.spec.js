import React from 'react';
import { shallow } from 'enzyme';
import { intlMock } from '@commercetools-local/test-utils';
import { UserProfileGeneralInfoPanel } from './user-profile-general-info-panel';

const createTestProps = props => ({
  isSubmitting: false,
  hasAttemptedSubmit: false,
  values: {
    email: 'john@snow.got',
    firstName: 'John',
    lastName: 'Snow',
  },
  errors: {},
  intl: intlMock,
  touched: {},
  onChange: jest.fn(),
  onBlur: jest.fn(),
  ...props,
});

describe('rendering', () => {
  describe('when there are no errors', () => {
    let props;
    let wrapper;
    beforeEach(() => {
      props = createTestProps();
      wrapper = shallow(<UserProfileGeneralInfoPanel {...props} />);
    });
    it('should ensure layout structure', () => {
      expect(wrapper).toMatchSnapshot();
    });
    it('should render input field for firstName', () => {
      expect(wrapper).toRender({ name: 'firstName' });
    });
    it('should render input field for lastName', () => {
      expect(wrapper).toRender({ name: 'lastName' });
    });
    it('should render input field for email', () => {
      expect(wrapper).toRender({ name: 'email' });
    });
  });

  describe('when there is a validation error on first name', () => {
    let props;
    let wrapper;
    beforeEach(() => {
      props = createTestProps({
        errors: { firstNameMissing: true },
        touched: { firstName: true },
        hasAttemptedSubmit: true,
      });
      wrapper = shallow(<UserProfileGeneralInfoPanel {...props} />);
    });
    it('should ensure layout structure', () => {
      expect(wrapper).toMatchSnapshot();
    });
    it('should show the error', () => {
      expect(wrapper.find('FormattedMessage')).toHaveProp(
        'id',
        'Validation.required'
      );
    });
    it('should highlight the first name input', () => {
      expect(wrapper.find({ name: 'firstName' })).toHaveProp(
        'hasWarning',
        true
      );
    });
  });
  describe('when submitting', () => {
    let props;
    let wrapper;
    beforeEach(() => {
      props = createTestProps({
        isSubmitting: true,
        touched: { firstName: true },
      });
      wrapper = shallow(<UserProfileGeneralInfoPanel {...props} />);
    });
    it('should disable all input fields', () => {
      expect(wrapper).toMatchSnapshot();
    });
    it('should disable the first name input', () => {
      expect(wrapper.find({ name: 'firstName' })).toHaveProp(
        'isDisabled',
        true
      );
    });
  });
});
