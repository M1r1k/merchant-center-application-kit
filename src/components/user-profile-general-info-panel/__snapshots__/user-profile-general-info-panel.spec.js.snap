// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering should ensure layout structure 1`] = `
<CollapsiblePanel
  label="UserProfile.GeneralInfoPanel.title"
>
  <FormBox
    hasWarning={false}
  >
    <LabelField
      isRequired={true}
      title="UserProfile.GeneralInfoPanel.firstName"
    />
    <Field
      component={[Function]}
      name="firstName"
      shouldDisplayTooltip={[Function]}
      submitFailed={false}
    />
  </FormBox>
  <FormBox
    hasWarning={false}
  >
    <LabelField
      isRequired={true}
      title="UserProfile.GeneralInfoPanel.lastName"
    />
    <Field
      component={[Function]}
      name="lastName"
      shouldDisplayTooltip={[Function]}
      submitFailed={false}
    />
  </FormBox>
  <FormBox
    hasWarning={false}
  >
    <LabelField
      isRequired={false}
      title="UserProfile.GeneralInfoPanel.email"
    />
    <Field
      component={[Function]}
      disabled={true}
      name="email"
    />
  </FormBox>
</CollapsiblePanel>
`;
